tslearn/nonmyopic.py

```diff
@@ -0,0 +1,370 @@
+            Delta_k = (average_distance - d_k) / average_distance
+            proba_current_cluster = 1 / (
+                np.exp(self.lamb * minimum_distance_to_average)
+                + np.exp(self.lamb * (minimum_distance_to_average - Delta_k))
+                + 10e-6
+            )
+            sk.append(proba_current_cluster)
+            sum_sk = sum_sk + sk[-1]
+        final_sk = [(x / sum_sk) for x in sk]
+        return final_sk
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        for k in range(0, self.n_clusters):
+            cost_y = 0
+            for y in range(0, self.__n_classes_):
+                cost_y_hat = 0
+                for y_hat in range(0, self.__n_classes_):
+                    if y != y_hat:
+                        cost_y_hat = (
+                            cost_y_hat
+                            + self.pyhatyck_["pyhatycks{0}".format(k)][
+                                truncated_t + tau - self.minimum_time_stamp
+                            ][y][y_hat]
+                        )
+                cost_y = cost_y + self.pyck_[y, k] * cost_y_hat
+            cost = cost + proba_clusters[k] * cost_y
+        cost = cost + self.cost_time(t=truncated_t + tau)
+
+        return cost
+
+    @staticmethod
+    def minimize_integer(end_of_time, function_to_minimize, xt):
+        """
+        We want to minimize a function "funct" according a time series "xt" for a list of integers from 0 to
+        "end_of_time"
+
+        Parameters
+        ----------
+        end_of_time: int
+            the highest integer at which the function to be minimized is calculated
+        function_to_minimize: funct
+            the function of interest
+        xt: vector
+            the time series related to the function "funct"
+
+        Returns
+         ------
+         int : The integer in {0,...,stop} that minimizes "funct"
+         float : the so fat minimum cost
+        """
+        tau_star = 0
+        so_far_minimum_cost = function_to_minimize(xt, tau_star)
+        for tau in range(1, end_of_time):
+            current_cost = function_to_minimize(xt, tau)
+            if current_cost < so_far_minimum_cost:
+                so_far_minimum_cost = current_cost
+                tau_star = tau
+        return tau_star + xt.shape[-1], so_far_minimum_cost
+
+    def _predict_at_fixed_length(self, xt, cost_function):
```

Comment:
I insist that `cost_function` should not be passed as an argument to that function: the function is private so it should not be call from outside the class. Also, if one wants to change the cost function for this estimator, the correct way would probably be to inherit from this class and change the `_cost_time` function.

Reply:
I also think the name of this function is misleading. It is not performing classification at fixed length, but rather is performing early classification for a single time series. Its prototype should probably be something like:
`def _predict_single_series(self, Xi):`